plex/player.py

```diff
@@ -0,0 +1,186 @@
+
+from plex.api import Api
+from plex.constants import PLEX_PROTOCOL, PLEX_PLAYER_PLAYING, PLEX_PLAYER_PAUSED, PLEX_PLAYER_STOPPED
+from plex.server import Server
+
+from lib.utils import log, mediaProvider2str
+
+REPORTING_INTERVAL = 5 # seconds
+
+class Player(xbmc.Player):
+    def __init__(self):
+        '''Initializes the player'''
+        super(xbmc.Player, self).__init__()
+
+        self._providers = {}
+        self._monitor = xbmc.Monitor()
+        self._lock = threading.Lock()
+
+    def AddProvider(self, mediaProvider):
+        '''Adds a media provider to the player'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        self._providers[mediaProvider.getIdentifier()] = mediaProvider
+
+    def RemoveProvider(self, mediaProvider):
+        '''Removes the associated media provider'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        del self._providers[mediaProvider.getIdentifier()]
+
+    def onPlayBackStarted(self):
+        '''Triggered when xbmc.Player is started'''
+        self._reset()
+        self._file = self.getPlayingFile()
+
+    def onAVStarted(self):
+        '''Triggered when playback actually starts'''
+        self._startPlayback()
+        self._syncPlaybackState(time=0, state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackSeek(self, time, seekOffset):
+        '''Triggered when seeking.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackSeekChapter(self, chapter):
+        '''Triggered when seeking chapters.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackPaused(self):
+        '''Triggered when playback is paused.'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PAUSED)
+
+    def onPlayBackResumed(self):
+        '''Triggered when playback is resumed after a pause'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackStopped(self):
+        '''Triggered when playback is stopped'''
+        self._syncPlaybackState(state=PLEX_PLAYER_STOPPED)
+        self.onPlayBackEnded()
+
+    def onPlayBackEnded(self):
+        '''Triggered when playback ends. Resets the player state and inherently kills the reporting loop'''
+        self._reset()
+
+    def _startPlayback(self):
+        '''Identifies the item (if from Plex) and initializes the player state'''
+        if not self._file:
+            return
+
+        if not self.isPlayingVideo():
+            return
+
+        videoInfoTag = self.getVideoInfoTag()
+        if not videoInfoTag:
+            return
+
+        itemId = videoInfoTag.getUniqueID(PLEX_PROTOCOL)
+        if not itemId:
+            return
+
+        if not itemId.isdigit():
+            log('invalid item id {} (non digit). Kodi will not report playback state to Plex Media Server' \
+                .format(matchingItems[0]), xbmc.LOGERROR)
```

Comment:
Maybe log the specific media provider.

Reply:
This was a leftover.  It is a sane check to verify the item is a digit since the plex API uses it as key (I doubt id's can be non-digits anyway) and the `updateTimeline` call expects an `int`. In the hypothetic case it is not an integer we should return early and avoid the expensive operation of getting the provider and the importer. If that operation can be made cheap, I'll add it later to the log item.

Reply:
In fact I was not returning early (fixed).